src/mlpack/methods/ann/layer/concat_impl.hpp

```diff
@@ -19,38 +19,81 @@
 namespace mlpack {
 namespace ann /** Artificial Neural Network. */ {
 
-template<typename InputType, typename OutputType>
-ConcatType<InputType, OutputType>::ConcatType(
-    const bool run) :
+template<typename MatType>
+ConcatType<MatType>::ConcatType(
+    const size_t axis) :
+    MultiLayer<MatType>(),
+    axis(axis),
+    useAxis(true)
+{
+  // Nothing to do.
+}
+
+template<typename MatType>
+ConcatType<MatType>::ConcatType() :
+    MultiLayer<MatType>(),
     axis(0),
     useAxis(false)
 {
   // Nothing to do.
 }
 
-template<typename InputType, typename OutputType>
-ConcatType<InputType, OutputType>::ConcatType(
-    const size_t axis,
-    const bool run) :
-    axis(axis),
-    useAxis(true)
+template<typename MatType>
+ConcatType<MatType>::~ConcatType()
 {
-  // Nothing to do.
+  // Nothing to do: the child layer memory is already cleared by MultiLayer.
 }
 
-template<typename InputType, typename OutputType>
-ConcatType<InputType, OutputType>::~ConcatType()
+template<typename MatType>
+ConcatType<MatType>::ConcatType(const ConcatType& other) :
+    MultiLayer<MatType>(other),
+    axis(other.axis),
+    useAxis(other.useAxis)
 {
-  // Clear memory.
-  for (size_t i = 0; i < this->network.size(); ++i)
-    delete this->network[i];
+  // Nothing else to do.
+}
+
+template<typename MatType>
+ConcatType<MatType>::ConcatType(ConcatType&& other) :
+    MultiLayer<MatType>(std::move(other)),
+    axis(std::move(other.axis)),
+    useAxis(std::move(other.useAxis))
+{
+  // Nothing else to do.
+}
+
+template<typename MatType>
+ConcatType<MatType>& ConcatType<MatType>::operator=(const ConcatType& other)
+{
+  if (this != &other)
+  {
+    MultiLayer<MatType>::operator=(other);
+    axis = other.axis;
+    useAxis = other.useAxis;
+  }
+
+  return *this;
 }
 
-template<typename InputType, typename OutputType>
-void ConcatType<InputType, OutputType>::Forward(
-    const InputType& input, OutputType& output)
+template<typename MatType>
+ConcatType<MatType>& ConcatType<MatType>::operator=(ConcatType&& other)
 {
-  this->InitializeForwardPassMemory();
+  if (this != &other)
+  {
+    MultiLayer<MatType>::operator=(std::move(other));
+    axis = std::move(other.axis);
+    useAxis = std::move(other.useAxis);
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+void ConcatType<MatType>::Forward(const MatType& input, MatType& output)
+{
+  // The implementation of MultiLayer is fine: this will allocate a matrix that
+  // is able to hold each child layer's output.
+  this->InitializeForwardPassMemory(input.n_cols);
```

Comment:
@shubham1206agra here is where I've reused `MultiLayer`'s memory allocation.  It turns out for the forward and backward pass, `MultiLayer::InitializeForwardPassMemory()` and `MultiLayer::InitializeBackwardPassMemory()` is fine: for the forward pass, I just need to have an output matrix whose size is equal to each layer's output dimensions, and this is exactly what `InitializeForwardPassMemory()` already does.  For the backward pass, I just need a matrix for each delta that has the same size as each layer's input dimensions.  (I'm not sure yet about the gradient ... my debugging hasn't gotten that far :))

Reply:
https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/multi_layer_impl.hpp#L388
In this, you can see that input size is taken through the prev layer output, which is true for MultiLayer but not for the Concat layer. That's why I needed to override this method.